pennylane/ops/op_math/decompositions/solovay_kitaev.py

```diff
@@ -63,25 +63,60 @@ def _quaternion_transform(matrix):
     )
 
 
-def _contains_SU2(op_mat, ops_vecs, tol=1e-8):
+def _contains_SU2(op_mat, ops_vecs=None, kd_tree=None, tol=1e-8):
     r"""Checks if a given SU(2) matrix is contained in a list of quaternions for a given tolerance.
 
     Args:
         op_mat (TensorLike): SU(2) matrix for the operation to be searched
         op_vecs (list(TensorLike)): List of quaternion for the operations that makes the search space.
+        kd_tree (KDTree): KDTree object built from the list of quaternions. Default is ``None``.
         tol (float): Tolerance for the match to be considered ``True``.
 
     Returns:
-        Tuple(bool, TensorLike): A bool that shows whether an operation similar to the given operations
-        was found, and the quaternion representation of the searched operation.
+        Tuple(bool, TensorLike, int): A tuple including `True`/`False` for whether an operation similar to the
+        given operations was found, the quaternion representation of the searched operations, and its index in
+        the `op_vecs` or `kd_tree`.
     """
-    node_points = qml.math.array(ops_vecs)
     gate_points = qml.math.array([_quaternion_transform(op_mat)])
 
-    tree = KDTree(node_points)
-    dist = tree.query(gate_points, workers=-1)[0][0]
+    tree = kd_tree or KDTree(qml.math.array(ops_vecs))
+    dist, indx = tree.query(gate_points, workers=-1)
 
-    return (dist < tol, gate_points[0])
+    return (dist[0] < tol, gate_points[0], indx[0])
+
+
+def _prune_approximate_set(
+    approx_set_ids, approx_set_mat, approx_set_gph, approx_set_qat, approx_set_sum
+):
+    """Prune the approximate set for equivalent gate sequences with higher T-gate counts.
+
+    Args:
+        approx_set_ids (list[list[~pennylane.operation.Operation]]): List of gate sequences.
+        approx_set_mat (list[TensorLike]): List of SU(2) matrices.
+        approx_set_gph (list[float]): List of global phases.
+        approx_set_qat (list[TensorLike]): List of quaternion representations.
+        approx_set_sum (list[int]): List of numbers of the T and Adjoint(T) gates in the sequences.
```

Comment:
```suggestion
    Args:
        approx_set_ids (list[list[~pennylane.operation.Operation]]): list of gate sequences
        approx_set_mat (list[TensorLike]): list of SU(2) matrices
        approx_set_gph (list[float]): list of global phases
        approx_set_qat (list[TensorLike]): list of quaternion representations
        approx_set_sum (list[int]): list of numbers of the T and Adjoint(T) gates in the sequences
```

https://docs.pennylane.ai/en/stable/development/guide/documentation.html